server.c

```diff
@@ -687,6 +687,65 @@ server_zonestat_switch(struct nsd* nsd)
 }
 #endif /* USE_ZONE_STATS */
 
+#ifdef BIND8_STATS
+void
+server_stat_alloc(struct nsd* nsd)
+{
+	char tmpfile[256];
+	size_t sz = sizeof(struct nsdst) * nsd->child_count * 2;
+	uint8_t z = 0;
+
+	/* file name */
+	nsd->statfname = 0;
+	snprintf(tmpfile, sizeof(tmpfile), "%snsd-xfr-%d/nsd.%u.stat",
+		nsd->options->xfrdir, (int)getpid(), (unsigned)getpid());
+	nsd->statfname = region_strdup(nsd->region, tmpfile);
+
+	/* file descriptor */
+	nsd->statfd = open(nsd->statfname, O_CREAT|O_RDWR, 0600);
+	if(nsd->statfd == -1) {
+		log_msg(LOG_ERR, "cannot create %s: %s", nsd->statfname,
+			strerror(errno));
+		unlink(nsd->zonestatfname[0]);
+		unlink(nsd->zonestatfname[1]);
+		exit(1);
+	}
+
+#ifdef HAVE_MMAP
+	if(lseek(nsd->statfd, (off_t)sz-1, SEEK_SET) == -1) {
+		log_msg(LOG_ERR, "lseek %s: %s", nsd->statfname,
+			strerror(errno));
+		unlink(nsd->statfname);
+		unlink(nsd->zonestatfname[0]);
+		unlink(nsd->zonestatfname[1]);
```

Comment:
Maybe use a single error handler? Each failure case has the three exact same operations and then `exit(1)`. Maybe use a `goto` or function? Also, isn't it the case that if you unlink, but keep the file descriptor open, the file is gone from the file system, but the file descriptor remains valid? In that case, technically, a `close` should be done first (not that it matters after `exit(1)`).

Reply:
Fixed in https://github.com/NLnetLabs/nsd/pull/292/commits/9557fd283e1a748f206ec534fa0efbd3774eb5d5 . This closes the statfd.